File: ReverseLinkedList/II/ReverseLinkedListII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        if head is None:
            return head
        pre_head = ListNode(None)
        pre_head.next = head
        pre = pre_head
        for i in range(m - 1):
            pre = pre.next
        start = pre.next
        next_start = start.next
        for i in range(n - m):
            start.next = next_start.next
            next_start.next = pre.next
            pre.next = next_start
            next_start = start.next
        return pre_head.next

File: ReverseLinkedList/II/ReverseLinkedListII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if(head == nullptr)
        	return head;
        ListNode* pre_head = new ListNode(0);
    	pre_head->next = head;
    	ListNode* pre = pre_head;
    	for(int i = 0; i < m - 1; i++) {
    		pre = pre->next;
    	}
    	ListNode* start = pre->next;
    	ListNode* next_start = start->next;
    	for(int i = 0; i < n - m; i++) {
    		start->next = next_start->next;
    		next_start->next = pre->next;
    		pre->next = next_start;
    		next_start = start->next;
    	}
    	return pre_head->next;
    }	
};